src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs

```diff
@@ -386,34 +410,156 @@ public override int GetHashCode()
+            var finalArray = builder.Pop();
+            finalArray = finalArray.WithElementType(finalArray.ElementTypeWithAnnotations.SetObliviousNullabilityForReferenceTypes());
+
+            // Unwind the stack and rebuild the array chain.
+            while (builder.Count > 0)
+            {
+                var state = builder.Pop();
+                var elementType = TypeWithAnnotations.Create(finalArray, NullableAnnotation.Oblivious, state.ElementTypeWithAnnotations.CustomModifiers);
+                finalArray = state.WithElementType(elementType);
+            }
+
+            builder.Free();
+            return finalArray;
         }
 
         internal override TypeSymbol MergeNullability(TypeSymbol other, VarianceKind variance)
         {
             Debug.Assert(this.Equals(other, TypeCompareKind.IgnoreDynamicAndTupleNames | TypeCompareKind.IgnoreNullableModifiersForReferenceTypes));
-            TypeWithAnnotations elementType = ElementTypeWithAnnotations.MergeNullability(((ArrayTypeSymbol)other).ElementTypeWithAnnotations, variance);
-            return WithElementType(elementType);
+            var builder = ArrayBuilder<(ArrayTypeSymbol thisArray, ArrayTypeSymbol otherArray)>.GetInstance();
+
+            // First build up the pair of arrays that need to be processed.
+            var state = (thisArray: this, otherArray: (ArrayTypeSymbol)other);
+            builder.Push(state);
+            do
+            {
+                var nextArray = state.thisArray.ElementType as ArrayTypeSymbol;
+                if (nextArray is null)
+                {
+                    break;
+                }
+                else
+                {
+                    state = (nextArray, (ArrayTypeSymbol)state.otherArray.ElementType);
+                    builder.Push(state);
+                }
+            }
+            while (true);
+
+            // Next build up the most nested element.
+            var lastState = builder.Pop();
+            var lastElementType = lastState.thisArray.ElementTypeWithAnnotations.MergeNullability(
+                lastState.otherArray.ElementTypeWithAnnotations,
+                variance);
+            var finalArray = lastState.thisArray.WithElementType(lastElementType);
+
+            // Unwind the stack and rebuild the array chain.
+            while (builder.Count > 0)
+            {
+                state = builder.Pop();
+                var nullableAnnotation = NullableAnnotationExtensions.MergeNullableAnnotation(
+                    state.thisArray.ElementTypeWithAnnotations.NullableAnnotation,
+                    state.otherArray.ElementTypeWithAnnotations.NullableAnnotation,
+                    variance);
+                var elementType = TypeWithAnnotations.Create(finalArray, nullableAnnotation, state.thisArray.ElementTypeWithAnnotations.CustomModifiers);
+                finalArray = state.thisArray.WithElementType(elementType);
+            }
+
+            return finalArray;
         }
```

Comment:
Notice I have a missing `builder.Free` call here. Everyone missed it on the review. I only noticed it as I was applying the refactorings suggested. 

This is what worries me about the separate implementations. Subtle issues like this.